### PR TITLE
Wazo 2185 linefeatures search

### PR DESCRIPTION
**Pull request by ja0x**

### CODE REVIEW

**Comment by fblackburn1** [comment]
rebase with master to include new zuul linters check

File: xivo_dao/alchemy/tests/test_linefeatures.py [hunk 1]
```diff
@@ -55,5 +56,50 @@ class TestApplication(DAOTestCase):
...
+
+        [line.append(self.add_line(context=context.name, endpoint_sip_uuid=sip_.uuid)) for sip_ in sip]
+        [line.append(self.add_line(context=context.name, endpoint_sccp_id=sccp_.id)) for sccp_ in sccp]
+
+        self._run_query_test(line, tests_count, lambda id: LineFeatures.caller_id_name == self._caller_id_name(id))
+        self._run_query_test(line, tests_count, lambda id: LineFeatures.caller_id_num == self._caller_id_num(id))
+        self._run_query_test(line, tests_count, lambda id: and_(
+            LineFeatures.caller_id_name == self._caller_id_name(id),
+            LineFeatures.caller_id_num == self._caller_id_num(id)
+        ))
```
Comment (fblackburn1) [inline xivo_dao/alchemy/tests/test_linefeatures.py:105]: Each test should test ~one thing
Split test for each endpoint
Moreover tests should remain simple and explicit and maintainable by junior/senior. There is no difference between testing 10 lines or 2. Only a lot of loop and unnecessary complexity are added. Don't hesitate to hardcode value to improve readability (only for tests...)

File: xivo_dao/resources/line/search.py [hunk 2]
```diff
@@ -15,7 +19,37 @@
...
+                               'exten': Extension.exten},
                       default_sort='name')
 
-line_search = SearchSystem(config)
+
+class LineSearchSystem(SearchSystem):
+
+    def search_from_query(self, query, parameters):
+        if 'exten' in parameters and isinstance(parameters['exten'], (str, six.text_type)):
+            extens = parameters.pop('exten').split(',')
```
Comment (fblackburn1) [inline xivo_dao/resources/line/search.py:33]: There is a reason to search by mutli exten? Ticket doesn't mention that
search example by single exten: https://github.com/wazo-platform/xivo-dao/commit/88a371e12691d810b402b8729b7d60b36b56b6ae

**Review by fblackburn1 — CHANGES_REQUESTED** [review]
## Blockers
* You add a lot of unused code (dead code) with `can_inherit=True`. It's not in our practice to do this kind of work. It add complexity to read/debug and maintains.
  * Moreover, tests are missing on this part too, nobody except you know if it work or not  
* unit tests are missing for the endpoint_sip `query_options_value` method. They should cover all corner cases scenarios.
* 
## Warnings
* be careful with your commit message length, you can read guideline in the wiki about max length
* You can rebase your branch **before** moving to review. It will avoid useless commit and be easier to validate. I can show you how, if you want.

## Comments
Do you have benchmark to compare before/after?
  * With > 1000 lines/endpoint_sip
  * With 1 line/endpoint_sip
Did you confirm this behavior with a PO (i.e. do not search through inherited template)?

File: xivo_dao/alchemy/endpoint_sip.py [hunk 3]
```diff
@@ -410,3 +418,82 @@ def _find_first_value(self, section, key):
...
+
+    @classmethod
+    def query_options_value(
+        cls,
+        option,
+        root_uuid=None,
+        sections=['endpoint'],
+        can_inherit=False
+    ):
+        check_sections = sections if isinstance(sections, list) else [sections]
```
Comment (fblackburn1) [inline xivo_dao/alchemy/endpoint_sip.py:430]: I don't see the point to search an option in many sections. There is no test and use case in the PR too

**Comment by rgarrigue** [comment]
recheck

**Review by fblackburn1 — CHANGES_REQUESTED** [review]
As discussed
* A documentation of small description of materialized with the PJSIP context would help to understand this part of code
* Check if you can use sqlalchemy-utils view
* Also, when you adapt external code, do not forget to add credit
  ex: `# Adapted from http://...`

**Review by fblackburn1 — APPROVED** [review]
:tada:
